arraykit.c

```diff
@@ -251,6 +261,716 @@ resolve_dtype_iter(PyObject *Py_UNUSED(m), PyObject *arg)
+        AK_GOTO_ON_NOT(other_unique, failure)
+        Py_INCREF(other_unique);
+
+        // 3. Concatenate
+        concatenated = AK_concat_arrays(raveled_array_unique, other_unique);
+        Py_DECREF(arr_and_rev_idx);
+        Py_DECREF(raveled_array_unique);
+        Py_DECREF(other_unique);
+    }
+    else {
+        // 3. Concatenate
+        concatenated = AK_concat_arrays(flattened_array, other);
+    }
+    AK_GOTO_ON_NOT(concatenated, failure)
+
+    size_t concatenated_size = PyArray_SIZE(concatenated);
+
+    // 4: Sort
+    ordered_idx = (PyArrayObject*)PyArray_ArgSort(concatenated, 0, NPY_MERGESORT);
+    AK_GOTO_ON_NOT(ordered_idx, failure)
+    npy_intp* ordered_idx_arr = (npy_intp*)PyArray_DATA(ordered_idx);
+
+    // 5. Find duplicates
+    sorted_arr = (PyArrayObject*)PyObject_GetItem((PyObject*)concatenated, (PyObject*)ordered_idx);
+    AK_GOTO_ON_NOT(sorted_arr, failure)
+
+    comparison = AK_compare_two_slices_from_array(sorted_arr, 1, concatenated_size, 0, concatenated_size - 1, Py_EQ);
+    AK_GOTO_ON_NOT(comparison, failure)
+    npy_bool* comparison_arr = (npy_bool*)PyArray_DATA(comparison);
+
+    if (!assume_unique) {
+        // 6: Construct empty array
+        PyObject* tmp = PyArray_Empty(
+                PyArray_NDIM(concatenated),      // nd
+                PyArray_DIMS(concatenated),      // dims
+                PyArray_DescrFromType(NPY_BOOL), // dtype
+                0);                              // is_f_order
+
+        Py_INCREF(tmp);
+
+        npy_intp dims[1] = {1};
+
+        PyArrayObject *false = (PyArrayObject*)PyArray_NewFromDescr(
+                &PyArray_Type,                         // class (subtype)
+                PyArray_DescrFromType(NPY_BOOL),       // dtype (descr)
+                1,                                     // ndim (nd)
+                dims,                                  // dims
+                NULL,                                  // strides
+                "\0",                                  // data
+                NPY_ARRAY_DEFAULT | NPY_ARRAY_OWNDATA, // flags
+                NULL);                                 // sublclass (obj)
+        Py_INCREF(false);
+
+        PyArrayObject* xyz = AK_concat_arrays(comparison, false);
+        Py_INCREF(xyz);
+
+        // TODO: Comparison is missing a trailing False value...
+        if (PyObject_SetItem(tmp, (PyObject*)ordered_idx, (PyObject*)xyz)) {
+            goto failure;
+        }
+
+        Py_INCREF(ordered_idx);
+        Py_INCREF(xyz);
+        Py_INCREF(tmp);
+        Py_INCREF(reverse_idx);
+
+        ret = (PyArrayObject*)PyObject_GetItem(tmp, reverse_idx);
+
+        if (array_ndim == 2) {
+            PyObject* shape = PyTuple_Pack(2, PyLong_FromLong(array_dims[0]), PyLong_FromLong(array_dims[1]));
+            ret = (PyArrayObject*)PyArray_Reshape(ret, shape);
+        }
+
+        Py_DECREF(tmp);
+        Py_DECREF(reverse_idx);
```

Comment:
You leak `false` and `xyz`, which should also be `DECREF`'d here. Although many of these can be `DECREF`'d much sooner above.